[clangd] Allow "move function body out-of-line" in non-header files

ckandeler opened the pull request:
Moving the body of member functions out-of-line makes sense for classes defined in implementation files too.

ckandeler committed:
[clangd] Allow "move function body out-of-line" in non-header files

Moving the body of member functions out-of-line makes sense for classes
defined in implementation files too.

ckandeler commented:
ping

kadircet commented:
Thanks! I believe the idea is great, but I am not really sure if this is useful enough without handling anon namespaces properly. It's very rare that someone has a class/struct in a cc file, that isn't in an anon-namespace.

Would you mind evolving the patch (and the infra) to work for method-definitions inside anon namespaces?

ckandeler commented:
> Would you mind evolving the patch (and the infra) to work for method-definitions inside anon namespaces?

Of course not; I didn't even now it wasn't working. 
I suppose the main point is to determine the insertion location such that it is still within the original namespace?
(I haven't looked at that part of the code so far.)

kadircet commented:
Yes, the current logic that finds insertion point wont work in presence of anon namespaces.

We normally do some pseudo parsing to find a location to insert the definiton, but because these are happening inside the main-file now, you can directly make use of sourcelocations inside the AST to determine an insertion point.

The other part is spelling of the function name and return type, there's a good chance they'll also do a weird thing when used for entities inside anon namespaces.

ckandeler commented:
The new patch set inserts at the end of the namespace block for the same-file case.

kadircet commented on clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp:
```diff
@@ -435,14 +407,17 @@ class DefineOutline : public Tweak {
 
-      // The refactoring is meaningless for unnamed classes and definitions
-      // within unnamed namespaces.
+      // The refactoring is meaningless for unnamed classes.
```
`... and namespaces, unless we're outlining in the same file`

kadircet commented on clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp:
```diff
@@ -435,14 +407,17 @@ class DefineOutline : public Tweak {
             return false;
         }
       }
+    } else if (SameFile) {
```
nit: might be better to move this to the top, e.g:
```
auto *MD = dyn_cast...;
if (!MD)
  return !SameFile; // Can't outline free-standing functions in the same file.

// rest of the verification for methods
```

kadircet commented on clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp:
```diff
@@ -499,17 +473,64 @@ class DefineOutline : public Tweak {
+    // we should (and, in the case of anonymous namespaces, need to)
+    // put the definition into the original namespace block.
+    // Within this constraint, the same considerations apply as in
+    // the FIXME below.
```
can you move fixme to the function comment instead?

kadircet commented on clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp:
```diff
@@ -499,17 +473,64 @@ class DefineOutline : public Tweak {
+    // Within this constraint, the same considerations apply as in
+    // the FIXME below.
+    if (SameFile) {
+      if (auto *Namespace = Source->getEnclosingNamespaceContext()) {
```
i think we should just fail the code action if enclosing namespace is nullptr, something is definitely wrong with that file (and falling back to textual matching feels suspicious).

kadircet replied:
yes, `getEnclosingNamespaceContext` can return `TranslationUnitDecl`, e.g. for:
```
struct Foo { void b^ar() {} };
```

and things are getting messy in that scenario, I missed the part around getting source locations, it'll actually be invalid in that case.

so what about changing this logic to get to enclosing decl instead (i.e. TagDecl) and insert right after its end location?

kadircet commented on clang-tools-extra/clangd/unittests/tweaks/DefineOutlineTests.cpp:
```diff
@@ -349,6 +349,44 @@ TEST_F(DefineOutlineTest, ApplyTest) {
+      // Member function with some adornments
+      // FIXME: What's with the extra spaces?
+      {
+          "namespace {\n"
```
can you use rawstring literals instead?

kadircet replied:
no need for all the complicated test cases, let's just validate the insertion location here for a test case like:
foo.cc
```
namespace foo {
namespace {
struct Foo { void ba^r() {} };
}
}
```

kadircet commented on clang-tools-extra/clangd/unittests/tweaks/DefineOutlineTests.cpp:
```diff
@@ -19,7 +19,7 @@ TWEAK_TEST(DefineOutline);
 TEST_F(DefineOutlineTest, TriggersOnFunctionDecl) {
   FileName = "Test.cpp";
-  // Not available unless in a header file.
+  // Not available for free function unless in a header file.
```
can you add some availability tests here?

ckandeler replied:
Do you have suggestions? anon namespace in header is already covered.

kadircet replied:
well, available for:
```
namespace {
struct Foo { void ba^r() {} };
}
```
and
```
namespace ns {
struct Foo { void ba^r() {} };
}
```
and
```
struct Foo { void ba^r() {} };
```

unavailable for:
```
namespace ns {
struct Foo { void bar(); };
void Foo::b^ar() {}
}
```

kadircet reviewed (commented):
thanks, LG in general just some small adjustments

kadircet commented on clang-tools-extra/clangd/unittests/tweaks/DefineOutlineTests.cpp:
```diff
@@ -349,6 +358,36 @@ TEST_F(DefineOutlineTest, ApplyTest) {
+          R"cpp(
+            namespace foo {
+            namespace {
+            struct Foo { void ba^r() {} };
```
might be worth also inserting some extra declarations in between to see how insertion point selection works

ckandeler replied:
> so what about changing this logic to get to enclosing decl instead (i.e. TagDecl) and insert right after its end location?

That will break with nested classes, won't it?

ckandeler replied:
Ah, maybe getOuterLexicalRecordContext().

ckandeler replied:
Hm, the SourceLocations are weird: getEndLoc() points to the closing brace of the class. Do I really have to look for the semicolon manually?

ckandeler commented:
ping

ckandeler commented:
ping

ckandeler commented:
ping

ckandeler commented:
I'm going to merge this without explicit format approval based on the following:
  - There was a general LGTM.
  - I addressed the remaining issues.
  - There was no further feedback for more than half a year.
